AdSec/Python/AdsecApiFactorised.py

```diff
@@ -0,0 +1,235 @@
+    i_load = ILoad.Create(Force(float(force), ForceUnit.Kilonewton),
+                          Moment(float(moment_yy), MomentUnit.KilonewtonMeter),
+                          Moment(float(moment_zz), MomentUnit.KilonewtonMeter))
+    uls_deformation = strength_results.Check(i_load).Deformation
+    sls_deformation = serviceability_results.Check(i_load).Deformation
+    print("\nULS deformation X " + str(uls_deformation.X.MilliStrain) +
+          " YY: " + str(uls_deformation.YY.PerMillimeters) +
+          " zz: " + str(uls_deformation.ZZ.PerMillimeters))
+    uls_load_util = strength_results.Check(i_load).LoadUtilisation
+    print("ULS utilisation: " + str(round(uls_load_util.Percent, 2)) + "%")
+
+
+def section_stress_strain():
+    global section
+    global design_code
+    global uls_deformation
+    global sls_deformation
+    print("\n\nStress - strain at key points of component")
+    flat_section = IAdSec.Create(design_code).Flatten(section)
+    corner_points = IPerimeterProfile(flat_section.Profile).SolidPolygon.Points
+    for points in corner_points:
+        point = IPoint(points)
+        y = point.Y.Millimeters
+        z = point.Z.Millimeters
+        strain = uls_deformation.StrainAt(point)
+        stress = flat_section.Material.Strength.StressAt(strain)
+        print("\nPoint: (" + str(y) + "," + str(z) + ")")
+        print("ULS Stress: " + str(round(stress.Megapascals, 2)) +
+              " Strain: " + str(round(strain.MilliStrain, 2)))
+        strain = sls_deformation.StrainAt(point)
+        stress = flat_section.Material.Strength.StressAt(strain)
+        print("SLS Stress: " + str(round(stress.Megapascals, 2)) +
+              " Strain: " + str(round(strain.MilliStrain, 2)))
+
+
+def rebar_stress_strain():
+    global section
+    global design_code
+    global uls_deformation
+    global sls_deformation
+    print("\n\nULS Stress - strain in rebar")
+    flat_section = IAdSec.Create(design_code).Flatten(section)
+    for bar_group in range(1, len(section.ReinforcementGroups)):
+        rebar_positions = ISingleBars(
+            flat_section.ReinforcementGroups[bar_group]).Positions
+        for rebar in rebar_positions:
+            position = IPoint(rebar)
+            y = rebar.Y.Millimeters
+            z = rebar.Z.Millimeters
+            strain = uls_deformation.StrainAt(position)
+            stress = ISingleBars(flat_section.ReinforcementGroups[1]
+                                 ).BarBundle.Material.Strength.StressAt(strain)
+            print("Point: (" + str(round(y, 1)) + "," + str(round(z, 1)) +
+                  ") :" +
+                  " Stress: " + str(round(stress.Megapascals, 1)) +
+                  " Strain: " + str(round(strain.MilliStrain, 2)))
+
+
+if __name__ == "__main__":
+    print("\nThis AdSec API version is " + IVersion.Api())
+    print("____")
+    
+    code_ec2_gb()
+
+    new_rectangle_section(700, 500, 30)
+    add_links(12)
```

Comment:
I had expected links would **not** be added to the rectangular section, since  `global section`  was not specified within add_links(), but post line 221, section did have links (I am confused!) 

```
def add_links(diameter):
    global bar_material
    bar_dia = Length(float(diameter), LengthUnit.Millimeter)
    rebar = ILinkGroup.Create(IBarBundle.Create(bar_material, bar_dia))
    **section**.ReinforcementGroups.Add(rebar)
```
My understanding is that any operation on 'section' inside a function would be a local change within that function and shouldn't be reflected on the global section. Ideally I would need to declare global section within add_links() in order for the change to be operated on the global variable. But this isn't what has happened in this script. Same applies to all the rebar types added.  Am I missing something? 
![image](https://user-images.githubusercontent.com/77726495/186396566-1553d00c-b125-4f4f-9fab-23c8ad0c2f3d.png)

Reply:
@VaradaNambiar I do not understand either why the functions are working without the global definition of the section. My suspicion is that the API is not respecting the local function restrictions in some way, which could be a concern.
@tim-lyon-arup @daviddekoning @nathanrobertsarup do you have any thoughts on this?

Just to be on the safe side I have added "global section" into each rebar function

Reply:
global variables are always accessible in lowe scopes in Python. The `global` keyword means that changes in the local scope will be reflected back up to the global scope.

If you just want to use a global variable in a lower scope, you *do not* need to explictly declare them as global

@PeterDebneyOasys @VaradaNambiar

Reply:
@daviddekoning in this instance, using the **section** in function changed the global instance, despite the function not declaring it as **global**, hence the confusion. So the following code does add rebar to the section. Is that as expected?

def add_links(diameter):
    global bar_material
    bar_dia = Length(float(diameter), LengthUnit.Millimeter)
    rebar = ILinkGroup.Create(IBarBundle.Create(bar_material, bar_dia))
    section.ReinforcementGroups.Add(rebar)

@VaradaNambiar

Reply:
hi @PeterDebneyOasys @VaradaNambiar , my apologies, I lost track of my github mentions. Did you manage to sort out this issue?